src/main/java/com/squareup/javapoet/CodeWriter.java

```diff
@@ -320,18 +325,31 @@ private void importableType(ClassName className) {
     }
   }
 
+  ClassName resolve(String simpleName, TypeSpec typeSpec, int i) {
+    if (Objects.equals(typeSpec.name, simpleName)) {
+      return stackClassName(i, simpleName);
+    }
+    for (TypeSpec nested : typeSpec.typeSpecs) {
+      if (Objects.equals(nested.name, simpleName)) {
+        return stackClassName(i, simpleName);
+      }
+      return resolve(simpleName, nested, i);
```

Comment:
`i` needs some love, shift, side-step ... before it is passed to the recursive call. As is, it produces slightly different ClassNames, but everything is compilable.